File: Arrays/1. Reversing an Array/revArray.cpp
#include<iostream>
using namespace std;
void printArray(int arr[], int n){
    for(int i = 0; i < n; i++){
        cout<<arr[i]<<' ';
    }
}
void iterative(int arr[], int n){
    int start = 0, end = n - 1, temp;
    while(start < end){
        temp = arr[start];
        arr[start] = arr[end];
        arr[end] = temp;

        start++;
        end--;
    }
}
void recursive(int arr[], int start, int end){
    if(start >= end){
        return;
    }

    int temp = arr[start];
    arr[start] = arr[end];
    arr[end] = temp;

    return recursive(arr, start+1, end-1);
}
int main(){
    int a[5] = {2,4,3,5,6};
    int n = sizeof(a)/sizeof(a[0]);
    cout <<"Before\n";
    printArray(a, n);
    recursive(a, 0, n-1);
    cout <<"\nAfter\n";
    printArray(a, n);
    return 0;
}

File: Arrays/1. Reversing an Array/RevArray.java
import java.io.*;

public class RevArray {

    static void printArr(int a[], int n){
        for(int i = 0; i < n; i++){
            System.out.print(a[i] + " ");
        }
        System.out.println();
    }
   
    // 1. Iterative method - O(n)
    //      i) Initialize start = 0, end = n-1
    //      ii) In loop, 
    //              Swap a[start] & a[end]
    //              Change start = start + 1, end = end -1

    static void iterative(int a[], int start, int end){
        int temp;

        while(start < end){
            temp = a[start];
            a[start] = a[end];
            a[end] = temp;

            start++;
            end--;
        }
    }

    // 2. Recursive method - O(n)
    //      i) Initialize start = 0, end = n-1
    //      ii) Swap a[start] & a[end]
    //      iii) Call recursively for rest of the array with start+1, end-1

    static void recursive(int a[], int start, int end){
        int temp;
        if(start >= end){
            return;
        }

        temp = a[start];
        a[start] = a[end];
        a[end] = temp;

        recursive(a, start+1, end-1);
    }
 
    public static void main(String args[]){

        int arr[] = {1,2,5,3,6,8};
        int n = 6;
        System.out.print("Before - ");
        printArr(arr, n);
        //iterative(arr, 0, n-1);
        recursive(arr, 0, n-1);
        System.out.print("After  - ");
        printArr(arr, n);
    }
}